### PR TITLE
feat: add type ascription

### PR DESCRIPTION
**Pull request by conradludgate**
Based on this conversation: https://twitter.com/rust_analyzer/status/1366092401278922757
Built off of `add_turbo_fish`, finds the current `let` statement, checks if it has type/turbofish already and checks if the rhs function is generic.

There's one case I couldn't figure out how to implement that would be nice:

```rust
#[test]
fn add_type_ascription_function_result() {
    check_assist(
        add_type_ascription,
        r#"
fn make<T>() -> Result<T, &'static str> {}
fn main() {
    let x = make()$0;
}
"#,
        r#"
fn make<T>() -> Result<T, &'static str> {}
fn main() {
    let x: Result<${0:_}, &'static str> = make();
}
"#,
    );
}
```

The `Function::ret_type` fn wasn't returning anything much useful so I'm not sure how to identity such scenarios just yet

### CODE REVIEW

**Comment by matklad** [comment]
Would adding this directly to `add_turbo_fish` work?  Just before we do 

```rust
    acc.add(
        AssistId("add_turbo_fish", AssistKind::RefactorRewrite),
```

we can write the following 

```rust
if let Some(let_stmt) = some logic here {
    acc.add(
        AssistId("add_type_ascription", AssistKind::RefactorRewrite),
}
acc.add(
    AssistId("add_turbo_fish", AssistKind::RefactorRewrite),

```

**Comment by conradludgate** [comment]
@matklad That seems to work quite nicely, thanks for the idea

File: crates/ide_assists/src/handlers/add_turbo_fish.rs [hunk 1]
```diff
@@ -50,8 +52,27 @@ pub(crate) fn add_turbo_fish(acc: &mut Assists, ctx: &AssistContext) -> Option<(
     let generics = hir::GenericDef::Function(fun).params(ctx.sema.db);
     if generics.is_empty() {
         mark::hit!(add_turbo_fish_non_generic);
+        mark::hit!(add_type_ascription_non_generic);
```
Comment (matklad) [inline crates/ide_assists/src/handlers/add_turbo_fish.rs:55]: It doesn't make much sense to have more than one test for a single mark, one of these tests can be safely removed.

**Comment by matklad** [comment]
bors r+

Thanks!

**Comment by lnicola** [comment]
ascription

**Pull request merged**